datafusion/core/src/physical_plan/aggregates/priority_map.rs

```diff
@@ -0,0 +1,969 @@
+        };
+        let mut valid = true;
+        if let Some(hi) = hi {
+            let label = format!("val={:?} idx={}, bucket={}", hi.val, idx, hi.map_idx);
+            builder.begin_child(label);
+            valid &= self._tree_print(idx * 2 + 1, builder); // left
+            valid &= self._tree_print(idx * 2 + 2, builder); // right
+            builder.end_child();
+            if idx != 0 {
+                let parent_idx = (idx - 1) / 2;
+                let parent = self.heap[parent_idx].as_ref().expect("Missing parent");
+                if (!self.desc && hi.val.comp(&parent.val) == Ordering::Greater)
+                    || (self.desc && hi.val.comp(&parent.val) == Ordering::Less)
+                {
+                    return false;
+                }
+            }
+        } else {
+            builder.add_empty_child("None".to_string());
+        }
+        valid
+    }
+
+    #[allow(dead_code)]
+    pub fn tree_print(&self) {
+        let mut builder = ptree::TreeBuilder::new("BinaryHeap".to_string());
+        let valid = self._tree_print(0, &mut builder);
+        let mut actual = Vec::new();
+        ptree::write_tree(&builder.build(), &mut actual).unwrap();
+        println!("{}", String::from_utf8(actual).unwrap());
+        if !valid {
+            panic!("Heap invariant violated");
+        }
+    }
+}
+
+struct CustomMap<ID: KeyType> {
+    map: RawTable<MapItem<ID>>,
+    limit: usize,
+}
+
+impl<ID: KeyType> CustomMap<ID> {
+    pub fn new(limit: usize, capacity: usize) -> Self {
+        Self {
+            map: RawTable::with_capacity(capacity),
+            limit,
+        }
+    }
+
+    pub fn find(&self, hash: u64, mut eq: impl FnMut(&ID) -> bool) -> Option<usize> {
+        let bucket = self.map.find(hash, |mi| eq(&mi.id))?;
+        let idx = unsafe { self.map.bucket_index(&bucket) };
+        Some(idx)
+    }
+
+    pub unsafe fn heap_idx_at(&self, map_idx: usize) -> usize {
+        let bucket = unsafe { self.map.bucket(map_idx) };
+        unsafe { bucket.as_ref().heap_idx }
+    }
+
+    pub unsafe fn remove_if_full(&mut self, replace_idx: usize) -> usize {
+        if self.map.len() >= self.limit {
+            unsafe { self.map.erase(self.map.bucket(replace_idx)) };
+            0 // if full, always replace top node
+        } else {
+            self.map.len() // if we're not full, always append to end
+        }
+    }
+
+    unsafe fn update_heap_idx(&mut self, mapper: &[(usize, usize)]) {
+        for (m, h) in mapper {
+            unsafe { self.map.bucket(*m).as_mut().heap_idx = *h }
+        }
+    }
+
+    pub fn insert(
+        &mut self,
+        hash: u64,
+        id: ID,
+        heap_idx: usize,
+        mapper: &mut Vec<(usize, usize)>,
+    ) -> usize {
+        let mi = MapItem::new(hash, id, heap_idx);
+        let bucket = self.map.try_insert_no_grow(hash, mi);
+        let bucket = match bucket {
+            Ok(bucket) => bucket,
+            Err(new_item) => {
```

Comment:
We `insert_no_grow` so we can catch events that trigger re-indexing of buckets, so we can rebuild are indexes. Since `capacity = limit *10`, we have a low fill factor so this should happen infrequently. Even at `limit * 10` this is still considerably less memory than before, since limit will typically be many orders of magnitude less than the groupby cardinality.